### ISSUE TITLE
tests non-deterministic. was: test/index sometimes fails

### ISSUE THREAD

**Issue by dominictarr** [socketsupply/introducer#2]
```
while node test/index.js; do date; sleep 1; done
```

eventually fails...

**Comment by dominictarr** [comment]
it will be an ordering thing, i think

**Comment by dominictarr** [comment]
also test/reconnect

running in a loop:

```
C=0; while node $test ; do C=$((C+1)); echo passes: $C; done; echo $C runs
```

`test/reconnect.js` passes like 9/10 times...

```
# stay connected via keepalive
not ok 17 easy peer received from hard
  ---
    operator: ok
    expected: true
    actual:   false
    at: Test.<anonymous> (/home/dominic/c/introducer/test/reconnect.js:159:5)
    stack: |-
      Error: easy peer received from hard
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.<anonymous> (/home/dominic/c/introducer/test/reconnect.js:159:5)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
          at process.processImmediate (node:internal/timers:471:21)
  ...
not ok 18 hard peer received from easy
  ---
    operator: ok
    expected: true
    actual:   false
    at: Test.<anonymous> (/home/dominic/c/introducer/test/reconnect.js:160:5)
    stack: |-
      Error: hard peer received from easy
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.<anonymous> (/home/dominic/c/introducer/test/reconnect.js:160:5)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
          at process.processImmediate (node:internal/timers:471:21)

```

**Comment by dominictarr** [comment]
test/index just passed 95 times then failed like this:

```
not ok 3 should be truthy
  ---
    operator: ok
    expected: true
    actual:   undefined
    at: Test.<anonymous> (/home/dominic/c/introducer/test/index.js:60:5)
    stack: |-
      Error: should be truthy
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.<anonymous> (/home/dominic/c/introducer/test/index.js:60:5)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
          at process.processImmediate (node:internal/timers:471:21)
  ...
not ok 4 should be truthy
  ---
    operator: ok
    expected: true
    actual:   undefined
    at: Test.<anonymous> (/home/dominic/c/introducer/test/index.js:61:5)
    stack: |-
      Error: should be truthy
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.<anonymous> (/home/dominic/c/introducer/test/index.js:61:5)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
          at process.processImmediate (node:internal/timers:471:21)
  ...
# intro, easy nat
ok 5 should be strictly equal
ok 6 should be truthy
ok 7 should be truthy
# detect hard nat
ok 8 should be strictly equal

1..8
# tests 8
# pass  6
# fail  2

```

**Comment by dominictarr** [comment]
test/chat.js

```
not ok 14 should be truthy
  ---
    operator: ok
    expected: true
    actual:   undefined
    at: assert_connected (/home/dominic/c/introducer/test/chat.js:115:7)
    stack: |-
      Error: should be truthy
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at assert_connected (/home/dominic/c/introducer/test/chat.js:115:7)
          at Test.<anonymous> (/home/dominic/c/introducer/test/chat.js:121:3)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
  ...
/home/dominic/c/introducer/test/chat.js:116
    t.ok(Object.keys(connections).length)
                ^

TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at assert_connected (/home/dominic/c/introducer/test/chat.js:116:17)
    at Test.<anonymous> (/home/dominic/c/introducer/test/chat.js:121:3)
    at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
    at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
    at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
    at Immediate.next (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
    at process.processImmediate (node:internal/timers:471:21)

Node.js v18.7.0

```

**Comment by dominictarr** [comment]
test/restart passed 100 times

**Comment by dominictarr** [comment]
test/swarm passed 100 times

**Comment by dominictarr** [comment]
test/wakeup

93 runs then

```
not ok 6 easy has found hard again
  ---
    operator: ok
    expected: true
    actual:   undefined
    at: Test.<anonymous> (/home/dominic/c/introducer/test/wakeup.js:101:5)
    stack: |-
      Error: easy has found hard again
          at Test.assert [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:314:54)
          at Test.bound [as _assert] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.assert (/home/dominic/c/introducer/node_modules/tape/lib/test.js:433:10)
          at Test.bound [as ok] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.<anonymous> (/home/dominic/c/introducer/test/wakeup.js:101:5)
          at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
          at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
          at Immediate.next [as _onImmediate] (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
          at process.processImmediate (node:internal/timers:471:21)
  ...
ok 7 should be strictly equal
/home/dominic/c/introducer/test/wakeup.js:104
  t.equal(peer_hard.data[swarm].length, 1)
                                ^

TypeError: Cannot read properties of undefined (reading 'length')
    at Test.<anonymous> (/home/dominic/c/introducer/test/wakeup.js:104:33)
    at Test.bound [as _cb] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
    at Test.run (/home/dominic/c/introducer/node_modules/tape/lib/test.js:117:31)
    at Test.bound [as run] (/home/dominic/c/introducer/node_modules/tape/lib/test.js:99:32)
    at Immediate.next (/home/dominic/c/introducer/node_modules/tape/lib/results.js:88:19)
    at process.processImmediate (node:internal/timers:471:21)
```

**Comment by dominictarr** [comment]
that's all the non-randomized tests... 

that's less deterministic than I was expecting.
I will investigate what is causing this...

**Comment by dominictarr** [comment]
if these tests fail sometimes then they should be randomized tests, instead

**Comment by dominictarr** [comment]
test/index is just one peer attempts to connect to the other using `intro`, both peers have static nat... this is so simple it should definitely work. hard to imagine it's a message ordering thing!?

**Comment by dominictarr** [comment]
in test/reliable, the test checks that `on_peer` gets called. it doesn't in the failure case.

**Comment by dominictarr** [comment]
test/wakeup checks that the easy peer has found the hard peer. a connection thing again.

**Comment by dominictarr** [comment]
I ported test/wakeup to a randomized test, but it always passed!
so it must have been something else...
I discovered a `Date.now()` in index (that shouldn't be there) but that didn't fix it.

I overrode Math.random in the failing test to be a deterministic random but seeded with Date.now() and it still failed, and I can reseed and then it fails immediately, so I can reproduce, now!

I should probably refactor the random tests so that any tap test can be randomized... depending on what I discover

**Comment by dominictarr** [comment]
test/wakeup seems like it isn't rejoining the swarm correctly. maybe because it needs to reestablish it's nat or something?

**Comment by dominictarr** [comment]
idea: a way to add eventual assertions

**Comment by dominictarr** [comment]
okay, I figured out the problem in test/wakeup.
join uses the `this.swarms` object to check the number of current peers, but after a wakeup these peers are probably connectable (it may have been some time since you connected, or even you are on a different ip now, so it just needs to check that they are actually active)

**Comment by dominictarr** [comment]
there were other tests with non-deterministic failures, so I made a script to run all tests randomly and figure out which ones were broken

```
# !!! test/chat.js FAILED after 14 runs
SEED=1668768695946100637 node test/chat.js

# !!! test/index.js FAILED after 58 runs
SEED=1668768739179481550 node test/index.js

# !!! test/reconnect.js FAILED after 17 runs
SEED=1668769001348498682 node test/reconnect.js

# !!! test/reliable.js FAILED after 14 runs
SEED=1668769003255496578 node test/reliable.js

# !!! test/wakeup.js FAILED after 6 runs
SEED=1668769289344348864 node test/wakeup.js
```

the other tests passed 1000 runs

**Comment by dominictarr** [comment]
surprised to see wakeup back in the list because I thought I fixed that but maybe it's something else this time?

**Comment by dominictarr** [comment]
this time it was a order bug in the tests. coming out of sleep can trigger an event if it would have been triggered.
the fix was declaring the expected message before unsleeping

**Comment by dominictarr** [comment]
wakeup now passes 1000 times!

**Comment by dominictarr** [comment]
progress, test/chat was failing because although the peers were actually static, when you check a static nat, it looks easy for a moment (depending on message order) and then when a static peer is told to connect, to another peer that isn't static, it didn't do anything (because they'll definitely be able to connect, since we are static). It was actually connecting, but it wasn't _tracking_ the connection properly so the test was failing. Fixed by having static peers do a ping3 as if it's an easy peer for any other peer nat type

**Comment by dominictarr** [comment]
hmm oh everything with a easy/hard connection is gonna fail sometimes. if it gives up after 2000 messages there will aways be some random seed that chooses the wrong ports and it fails to make a connection. so these tests need to just keep sending packets...

**Comment by dominictarr** [comment]
changed how constants works so that it's possible to override bdpMaxPackets in the tests

**Comment by dominictarr** [comment]
looks like there is just one thing left... that test/reliable isn't calling on_peer.
I thought maybe if you join a swarm but you are actually already connect then it doesn't notify but.... it should, because the peer is new to the swarm.

**Comment by dominictarr** [comment]
on test/swarms there are some codepaths where introducers join the swarm. which they are not meant to. it's something to do with reconnecting or timeouts... it causes the peer to rejoin, and now it looks like the introducer wants to be part of the swarm, other peers suggesting connections to that introducer

**Comment by dominictarr** [comment]
fixed in 7.4.0
